### PR TITLE
Make index maintenance cron jobs configurable

### LINKED ISSUE #1404: Make the initial maven index related cron jobs cronExpressions configurable
# Task Description
When new maven repository is created (including initial strongbox startup with pre-configured repositories) and its indexing is enabled one of the following cron jobs [RebuildMavenIndexesCronJob](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/cron/jobs/RebuildMavenIndexesCronJob.java), [DownloadRemoteMavenIndexCronJob](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/cron/jobs/DownloadRemoteMavenIndexCronJob.java) and [MergeMavenGroupRepositoryIndexCronJob](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/cron/jobs/MergeMavenGroupRepositoryIndexCronJob.java) (depending on the repository type) is scheduled for this repository. By default `cronExpression`s for these jobs are hard coded (see [MavenRepositoryManagementStrategy.java#L37](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenRepositoryManagementStrategy.java#L37). These cron expressions should be configurable in [strongbox.yaml](https://github.com/strongbox/strongbox/blob/master/strongbox-resources/strongbox-storage-api-resources/src/main/resources/etc/conf/strongbox.yaml) (see `repositoryConfiguration`, it should be sibling `indexingEnabled`). See also [MavenRepositoryConfiguration](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/yaml/configuration/repository/MavenRepositoryConfiguration.java) and its implementations.

# Tasks

The following tasks will need to be carried out:
* [ ]  `String getCronExpression();` method should be added to [MavenRepositoryConfiguration](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/yaml/configuration/repository/MavenRepositoryConfiguration.java) interface. `cronExpression` field should be added to its implementations
* [ ]  `strongbox.yaml` should be improved to allow specify both `cronExpression` as described in task description
* [ ] above `cronExpression`s should be used in `MavenRepositoryManagementStrategy#createRemoteIndexDownloaderCronTask` and `MavenRepositoryManagementStrategy#createRebuildMavenIndexCronJob`
* [ ] REST API should be improved too. See [MavenRepositoryConfigurationForm](https://github.com/strongbox/strongbox/blob/master/strongbox-web-forms/src/main/java/org/carlspring/strongbox/forms/configuration/MavenRepositoryConfigurationForm.java) and [MavenRepositoryConfigurationConverter](https://github.com/strongbox/strongbox/blob/master/strongbox-web-core/src/main/java/org/carlspring/strongbox/converters/configuration/MavenRepositoryConfigurationConverter.java)

# Task Relationships

This task:
* Depends on: ~~https://github.com/strongbox/strongbox/issues/1202~~
* Is a follow-up of: ~~https://github.com/strongbox/strongbox/issues/1202~~

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @fuss86
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by cbrianhill**
# Pull Request Description

This pull request closes #1404 

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [ ] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [x] Yes, please see https://github.com/strongbox/strongbox-docs/pull/39
  * [ ] No

### CODE REVIEW

File: strongbox-resources/strongbox-storage-api-resources/src/main/resources/etc/conf/strongbox.yaml [hunk 1]
```diff
@@ -54,6 +54,8 @@ configuration:
           repositoryConfiguration:
             type: Maven 2
             indexingEnabled: true
+            downloadRemoteMavenIndexCronExpression: "0 0 2 * * ?"
+            rebuildMavenIndexesCronExpression: "0 0 2 * * ?"
```
Comment (fuss86) [inline strongbox-resources/strongbox-storage-api-resources/src/main/resources/etc/conf/strongbox.yaml]: actually we have 3 cron jobs 

`RebuildMavenIndexesCronJob` - works for hosted repository
`DownloadRemoteMavenIndexCronJob` - works for proxy repository
`MergeMavenGroupRepositoryIndexCronJob` - works for group repository

there should be one `***CronExpression` per repository depending on the repository type

File: strongbox-resources/strongbox-storage-api-resources/src/main/resources/etc/conf/strongbox.yaml [hunk 2]
```diff
@@ -65,6 +67,8 @@ configuration:
...
```
Comment (fuss86) [inline strongbox-resources/strongbox-storage-api-resources/src/main/resources/etc/conf/strongbox.yaml]: as above

File: strongbox-resources/strongbox-storage-api-resources/src/main/resources/etc/conf/strongbox.yaml [hunk 3]
```diff
@@ -82,6 +86,8 @@ configuration:
...
```
Comment (fuss86) [inline strongbox-resources/strongbox-storage-api-resources/src/main/resources/etc/conf/strongbox.yaml]: as above

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenRepositoryManagementStrategy.java [hunk 4]
```diff
@@ -45,14 +46,19 @@ protected void createRepositoryInternal(Storage storage,
...
+                                           repositoryId,
+                                           repositoryConfig.getRebuildMavenIndexesCronExpression());
         }
         if (repository.isProxyRepository())
         {
-            createRemoteIndexDownloaderCronTask(storageId, repositoryId);
+            createRemoteIndexDownloaderCronTask(storageId, repositoryId,
+                                                repositoryConfig.getDownloadRemoteMavenIndexCronExpression());
         }
         if (repository.isGroupRepository())
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenRepositoryManagementStrategy.java]: the same should be applied to group repository with indexing enabled

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/yaml/configuration/repository/MavenRepositoryConfiguration.java [hunk 5]
```diff
@@ -9,4 +9,8 @@
 
     boolean isIndexingClassNamesEnabled();
 
+    String getDownloadRemoteMavenIndexCronExpression();
+
+    String getRebuildMavenIndexesCronExpression();
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/yaml/configuration/repository/MavenRepositoryConfiguration.java]: missing for group repository

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/converters/configuration/MavenRepositoryConfigurationConverter.java [hunk 6]
```diff
@@ -17,6 +17,10 @@ public MavenRepositoryConfigurationDto convert(MavenRepositoryConfigurationForm
         MavenRepositoryConfigurationDto configuration = new MavenRepositoryConfigurationDto();
         configuration.setIndexingClassNamesEnabled(mavenRepositoryConfigurationForm.isIndexingClassNamesEnabled());
         configuration.setIndexingEnabled(mavenRepositoryConfigurationForm.isIndexingEnabled());
+        configuration.setDownloadRemoteMavenIndexCronExpression(
+                mavenRepositoryConfigurationForm.getDownloadRemoteMavenIndexCronExpression());
+        configuration.setRebuildMavenIndexesCronExpression(
+                mavenRepositoryConfigurationForm.getRebuildMavenIndexesCronExpression());
```
Comment (fuss86) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/converters/configuration/MavenRepositoryConfigurationConverter.java]: missing cron expression for `MergeMavenGroupRepositoryIndexCronJob `

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/yaml/configuration/repository/MavenRepositoryConfigurationDto.java [hunk 7]
```diff
@@ -12,13 +12,20 @@
...
 
     private boolean indexingEnabled = false;
 
     private boolean indexingClassNamesEnabled = true;
 
+    // defaults to once daily at midnight
+    private String downloadRemoteMavenIndexCronExpression = "0 0 0 * * ?";
+
+    // defaults to once daily at 2 am
+    private String rebuildMavenIndexesCronExpression = "0 0 2 * * ?";
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/yaml/configuration/repository/MavenRepositoryConfigurationDto.java]: missing cron expression for `MergeMavenGroupRepositoryIndexCronJob `

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java [hunk 8]
```diff
@@ -265,6 +265,8 @@ public void testAddGetRepository(String extension)
         MavenRepositoryConfigurationForm mavenRepositoryConfigurationForm = new MavenRepositoryConfigurationForm();
         mavenRepositoryConfigurationForm.setIndexingEnabled(true);
         mavenRepositoryConfigurationForm.setIndexingClassNamesEnabled(false);
+        mavenRepositoryConfigurationForm.setDownloadRemoteMavenIndexCronExpression("0 0 2 * * ?");
+        mavenRepositoryConfigurationForm.setRebuildMavenIndexesCronExpression("0 0 2 * * ?");
```
Comment (fuss86) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/configuration/StoragesConfigurationControllerTestIT.java]: missing cron expression for `MergeMavenGroupRepositoryIndexCronJob `

File: strongbox-web-forms/src/main/java/org/carlspring/strongbox/forms/configuration/MavenRepositoryConfigurationForm.java [hunk 9]
```diff
@@ -14,6 +14,10 @@
 
     private boolean indexingClassNamesEnabled;
 
+    private String downloadRemoteMavenIndexCronExpression;
+
+    private String rebuildMavenIndexesCronExpression;
```
Comment (fuss86) [inline strongbox-web-forms/src/main/java/org/carlspring/strongbox/forms/configuration/MavenRepositoryConfigurationForm.java]: missing cron expression for `MergeMavenGroupRepositoryIndexCronJob `

**Comment by fuss86** [comment]
@steve-todorov it will also affect the UI since we will be able to provide `cron expression` value for repository with indexing enabled

@carlspring do you think we should add 3 distinct cron expressions for each job ? Or should we just provide one property `cronExpression` and use this expression based on the repository type ? I think the latter is fine and easier since repository always has one from 3 available types

@cbrianhill thanks your PR :) Let's wait for @carlspring answer :)

**Comment by steve-todorov** [comment]
Thanks for the ping @fuss86!

> Do you think we should add 3 distinct cron expressions for each job?

I agree with your thoughts and this sounds reasonable to me. It should be easier to manage from a UI perspective as well. 

> @steve-todorov it will also affect the UI since we will be able to provide cron expression value for repository with indexing enabled

This is going to be a tricky one. From what I understand, if `indexingEnabled` is `true`, then the `cronExpression` needs to be required. If that's the case we'll need to think of a way to "provide" this information to the front end part so that it enables and possibly requires the field to be populated. Most likely we'll need to do this in a follow-up task.

**Comment by carlspring** [comment]
> @carlspring do you think we should add 3 distinct cron expressions for each job ? Or should we just provide one property `cronExpression` and use this expression based on the repository type ? I think the latter is fine and easier since repository always has one from 3 available types

In my opinion, these should be configurable separately. If you have many repositories with a large number of artifacts and if they're on the same physical hard drive, you will definitely not want to trigger all the cron tasks at the same time. There will be cases where you'd like the option to fine-tune this.

**Comment by carlspring** [comment]
Now that I've re-read the question, I see what you mean, @fuss86.

Yes, the name of the field should be `cronExpression`, but, just to clarify, this should be configurable for each of the cron jobs.

File: strongbox-web-forms/src/main/java/org/carlspring/strongbox/forms/configuration/MavenRepositoryConfigurationForm.java [hunk 10]
```diff
@@ -34,6 +38,20 @@ public void setIndexingClassNamesEnabled(final boolean indexingClassNamesEnabled
         this.indexingClassNamesEnabled = indexingClassNamesEnabled;
     }
 
+    public String getDownloadRemoteMavenIndexCronExpression() { return downloadRemoteMavenIndexCronExpression; }
```
Comment (carlspring) [inline strongbox-web-forms/src/main/java/org/carlspring/strongbox/forms/configuration/MavenRepositoryConfigurationForm.java]: Cosmetic: Please, don't go getters and setters on one line.

File: strongbox-web-forms/src/main/java/org/carlspring/strongbox/forms/configuration/MavenRepositoryConfigurationForm.java [hunk 11]
```diff
@@ -34,6 +38,20 @@ public void setIndexingClassNamesEnabled(final boolean indexingClassNamesEnabled
...
+
+    public void setDownloadRemoteMavenIndexCronExpression(String downloadRemoteMavenIndexCronExpression)
+    {
+        this.downloadRemoteMavenIndexCronExpression = downloadRemoteMavenIndexCronExpression;
+    }
+
+    public String getRebuildMavenIndexesCronExpression() { return rebuildMavenIndexesCronExpression; }
```
Comment (carlspring) [inline strongbox-web-forms/src/main/java/org/carlspring/strongbox/forms/configuration/MavenRepositoryConfigurationForm.java]: Cosmetic: Please, don't go getters and setters on one line.

**Comment by cbrianhill** [comment]
Thanks @carlspring, @fuss86, @steve-todorov for the feedback.  I will change this to use a single attribute called `cronExpression`, and the value will be applied to the correct cron job depending on the repository type as defined above.

**Comment by cbrianhill** [comment]
It appears the unit test failure in travis is one that is fixed with #1435.

**Comment by steve-todorov** [comment]
Thanks for your work on this @cbrianhill! Could you please rebase one more time? :)

**Comment by cbrianhill** [comment]
@steve-todorov for sure!

**Comment by steve-todorov** [comment]
Thanks for your work on this @cbrianhill and sorry for the delay in merging! :)

**Comment by fuss86** [comment]
Congrats @cbrianhill and thanks !

**Pull request merged**